### ISSUE TITLE
Potential timing analysis attack

### ISSUE THREAD

**Issue by hmil** [lbarman/vaultage#100]
@lbarman plz evaluate:  https://github.com/lbarman/vaultage/blob/d43dd54b37522e6943089df8e1aa72afadc07c7f/packages/vaultage/src/storage/JSONCipherRepository.ts#L72-L77

**Comment by lbarman** [comment]
Well, I don't think it is a big problem. We can first improve the security by doing 
``` const match1 = (contents.username !== creds.username)
const match2 = (contents.password !== creds.password)
if ( match1 ^ match2 === 0 )
```
Then, the fact that `!==` itself isn't constant... I'm not against randomizing the time by adding `M` comparisons of random `N` bits string before. A fake sleep, but realistic given that it takes the same amount of time (in mean) than the function we're trying to protect.

**Comment by lbarman** [comment]
As mentioned privately, typescript doesn't let you convert boolean to int, or do bitwise operators on booleans. Hence the simplest solution (but I'm no good in Typescript) would look like:
```
const mapper: {[key: string]: number} = {'true': 1, 'false': 0}
const match1 = (contents.username !== creds.username) as string
const match2 = (contents.password !== creds.password) as string
if ( mapper[match1] ^ mapper[match2] === 0 )
```
Which perhaps *is* more secure but *looks* shady and untrustworthy. Do you know of a better way ? I'm for keeping it as simple as possible if we don't have a clean, better solution.

**Comment by hmil** [comment]
Well as this is the Node.JS world, ofc the answer is... [There's a module for it!](https://www.npmjs.com/package/safe-compare)

**Comment by hmil** [comment]
JK the code is tiny, you could just copy/paste instead https://github.com/Bruce17/safe-compare/blob/master/index.js

**Comment by lbarman** [comment]
Fixed 9ce65b2c6a5dffb83e4f176cddabc6d02ba45964
